queue_processors/queue_processor/tests/test_queue_listener_daemon.py

```diff
@@ -0,0 +1,116 @@
+# Autoreduction Repository : https://github.com/ISISScientificComputing/autoreduce
+#
+# Copyright &copy; 2019 ISIS Rutherford Appleton Laboratory UKRI
+# SPDX - License - Identifier: GPL-3.0-or-later
+# ############################################################################### #
+"""
+Tests the Queue Listener Daemon
+"""
+
+import unittest
+from unittest import mock
+
+from queue_processors.queue_processor.queue_listener_daemon import \
+    QueueListenerDaemon
+import queue_processors.queue_processor.queue_listener_daemon
+from utils.clients.queue_client import QueueClient
+
+# pylint: disable=protected-access
+
+
+class TestQueueListenerDaemon(unittest.TestCase):
+    """
+    Tests both the class and main method for Queue Listener Daemon
+    """
+    @staticmethod
+    @mock.patch("queue_processors.queue_processor."
+                "queue_listener_daemon.QueueListenerDaemon")
+    @mock.patch("queue_processors.queue_processor."
+                "queue_listener_daemon.control_daemon_from_cli")
+    @mock.patch("logging.error")
+    def test_main(patched_logging, patched_cli, patched_daemon):
+        """
+        Tests the main method correctly starts and handles the nominal case
+        """
+        daemon_instance = patched_daemon.return_value
+        daemon_instance.safe_shutdown.wait.return_value = True
+        queue_processors.queue_processor.queue_listener_daemon.main()
+
+        patched_daemon.assert_called_once_with("/tmp/QueueListenerDaemon.pid")
+        patched_cli.assert_called_once_with(patched_daemon.return_value)
+        daemon_instance.safe_shutdown.wait.assert_called()
+        patched_logging.assert_not_called()
+
+    @staticmethod
+    @mock.patch("queue_processors.queue_processor."
+                "queue_listener_daemon.QueueListenerDaemon")
+    @mock.patch("logging.error")
+    def test_main_timeout_logs(patched_logging, patched_daemon):
+        """
+        Tests main will correctly log a timeout as an error
+        """
+        daemon_event = patched_daemon.return_value.safe_shutdown
+        daemon_event.wait.return_value = False  # i.e. unsafe shutdown
+        with mock.patch("queue_processors.queue_processor."
+                        "queue_listener_daemon.control_daemon_from_cli"):
+            queue_processors.queue_processor.queue_listener_daemon.main()
+
+        patched_logging.assert_called()
+
+    def setUp(self):
+        self.instance = QueueListenerDaemon(pidfile="/should/never/be/created")
+
+    @mock.patch("queue_processors.queue_processor."
+                "queue_listener_daemon.queue_processor")
+    @mock.patch("threading.Timer")
+    def test_run(self, patched_timer, patched_processor):
+        """
+        Tests that run method sets up the stop timer and the main worker thread
+        """
+        self.instance.run()
+
+        patched_processor.main.assert_called_once()
+        patched_timer.assert_called_once_with(mock.ANY, self.instance.stop)
```

Comment:
NITPICK: Can we always be certain that the type will be correct given mock.ANY doesn't perform type checking?

Reply:
True, but equally we have to start pulling apart call args and it violates the Python constraint EAFP

My thought process is:
If we pass in the wrong arg for the method invoked it's not immediately obvious and you need to wait 24 hours to find out
If you pass in the wrong arg (which could be say a `datetime` instead of an int) for the first param the Daemon will die before starting since the Timer class checks the types beforehand